code/modules/mob/living/simple_animal/hostile/headcrab.dm

```diff
@@ -196,3 +203,223 @@
+	. = ..()
+	home_nest_area = get_area(src)
+	home_nest_turf = get_turf(src)
+	do_gonarch_screech(8, 100, 8, 100)
+
+/mob/living/simple_animal/hostile/headcrab/gonarch/handle_automated_action()
+	. = ..()
+	if(!.)
+		return
+
+	if(home_nest_area == get_area(src))
+		mode_switch(MODE_REST)
+	else
+		homesick++
+
+	//combat
+	if(target && target != home_nest_turf && mode != MODE_RAMPAGE && mode != MODE_RETURNING) //We do not have a fight while rampage is on or while we return to base.
+		//we are in combat now, guaranteed.
+		//Is our Homesick too high to go rampage and leave?
+		if(homesick >= gonarch_rampage_trigger)
+			mode_switch(MODE_RAMPAGE, list('sound/creatures/gonarch_rampage.ogg'))
+		else
+			mode_switch(MODE_DEFENDING)
+		return
+
+	//out of combat
+	if(homesick >= gonarch_rampage_trigger && mode != MODE_RAMPAGE)
+		mode_switch(MODE_RAMPAGE, list('sound/creatures/gonarch_rampage.ogg'))
+		return
+	switch(mode)
+		if(MODE_REST)		// idle
+			gonarch_rest()
+		if(MODE_RETURNING)		// returning home
+			gonarch_return()
+		if(MODE_NEWHOME)		// Seeking new home
+			gonarch_newhome()
+		if(MODE_RAMPAGE)		// Cannot find nest, frustration at max - rampage
+			gonarch_rampage()
+		if(MODE_DEFENDING) //If it goes down here, there is no target anymore. And it should return.
+			mode_switch(MODE_RETURNING, list('sound/creatures/gonarch_return.ogg'))
+
+/mob/living/simple_animal/hostile/headcrab/gonarch/proc/mode_switch(mode_to_switch, sound_to_play)
+	if(mode_to_switch)
+		mode = mode_to_switch
+		if(sound_to_play)
+			var/sound_picked = pick(sound_to_play)
+			playsound(get_turf(src), sound_picked, 200, 1, 15, null, null, null, FALSE, TRUE)
+	else
+		stack_trace("[src] tried to switch to a NULL mode.")
+
+/mob/living/simple_animal/hostile/headcrab/gonarch/proc/gonarch_rest()
+	if(get_area(src) == home_nest_area) // If I am home & Everything is calm
+		if(aggro_vision_range != gonarch_aggro_vision_range || vision_range != gonarch_standard_vision_range || speed != gonarch_standard_speed)
+			aggro_vision_range = gonarch_aggro_vision_range
+			vision_range = gonarch_standard_vision_range
+			speed = gonarch_standard_speed
+		if(homesick)
+			homesick = 0
+			step_towards(src, home_nest_turf, speed)
+		make_headcrabs() //Its its own proc, so they can be called by admins.
+	else
+		if(homesick >= 5)
+			mode_switch(MODE_RETURNING, list('sound/creatures/gonarch_return.ogg'))
+
+/mob/living/simple_animal/hostile/headcrab/gonarch/proc/gonarch_return()
+	speed = -1
+	var/turf/olddist = get_dist(src, home_nest_turf)
+	step_towards(src, home_nest_turf, speed)
+	target = home_nest_turf
+	if((get_dist(src, target)) >= (olddist))
+		frustration++
+		if(frustration > frustration_limit)
+			DestroyPathToTarget()
+			homesick++
+	else
+		frustration = 0
+
+/mob/living/simple_animal/hostile/headcrab/gonarch/proc/gonarch_newhome()
+	var/list/turfs = new/list()
+	for(var/turf/T in range(src, gonarch_nestfind_range))
+		if(T in range(src, 4))
```

Comment:
You may wanna cache that `range(src, 4)` outside of the loop to avoid computing it for every turf